### PR TITLE
Add sla plugin for job resource reservation feature.

### PR DESCRIPTION
**Pull request by jiangkaihua**
Signed-off-by: jiangkaihua <jiangkaihua1@huawei.com>

### CODE REVIEW

File: pkg/scheduler/framework/session_plugins.go [hunk 1]
```diff
@@ -265,8 +265,14 @@ func (ssn *Session) JobPipelined(obj interface{}) bool {
 			}
 			hasFound = true
 
-			if !jrf(obj) {
+			res := jrf(obj)
+			switch res {
+			case 1:
```
Comment (k82cn) [inline pkg/scheduler/framework/session_plugins.go:270]: check whether less than 0 or bigger than 0, instaed of 1, -1.

File: pkg/scheduler/framework/session_plugins.go [hunk 2]
```diff
@@ -338,8 +344,14 @@ func (ssn *Session) JobEnqueueable(obj interface{}) bool {
 				continue
 			}
 			hasFound = true
-			if res := fn(obj); !res {
-				return res
+			res := fn(obj)
+			switch res {
+			case 1:
```
Comment (k82cn) [inline pkg/scheduler/framework/session_plugins.go:349]: ditto

File: pkg/scheduler/plugins/overcommit/overcommit.go [hunk 3]
```diff
@@ -94,26 +94,26 @@ func (op *overcommitPlugin) OnSessionOpen(ssn *framework.Session) {
...
-	ssn.AddJobEnqueueableFn(op.Name(), func(obj interface{}) bool {
+	ssn.AddJobEnqueueableFn(op.Name(), func(obj interface{}) int {
 		job := obj.(*api.JobInfo)
 		idle := op.idleResource
 		inqueue := api.EmptyResource()
 		inqueue.Add(op.inqueueResource)
 		if job.PodGroup.Spec.MinResources == nil {
 			klog.V(4).Infof("job <%s/%s> is bestEffort, allow it be inqueue", job.Namespace, job.Name)
-			return true
+			return 0
```
Comment (k82cn) [inline pkg/scheduler/plugins/overcommit/overcommit.go:104]: why ignore ?

File: pkg/scheduler/plugins/sla/sla.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2021 The Volcano Authors.
```
Comment (k82cn) [inline pkg/scheduler/plugins/sla/sla.go:2]: please align copyright with other file.

File: pkg/scheduler/plugins/sla/sla.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+	"volcano.sh/volcano/pkg/scheduler/api"
+	"volcano.sh/volcano/pkg/scheduler/framework"
+)
+
+const (
+	// PluginName indicates name of volcano scheduler plugin
+	PluginName = "sla"
+	// JobWaitingTime is maximum waiting time that a job could stay Pending in service level agreement
+	// when job waits longer than waiting time, it should be inqueue at once, and cluster should reserve resources for it
+	JobWaitingTime = "SLA-WaitingTime"
```
Comment (k82cn) [inline pkg/scheduler/plugins/sla/sla.go:30]: s/SLA-WaitingTime/sla-waiting-time

File: pkg/scheduler/plugins/sla/sla.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+// readJobWaitingTime read in job waiting time for job
+// TODO: should also read from given field in volcano job spec
+func (sp *slaPlugin) readJobWaitingTime(jobInfo *api.JobInfo) (time.Duration, error) {
+	// read individual jobInfo waiting time from jobInfo annotations
+	if _, exist := jobInfo.PodGroup.Annotations[JobWaitingTime]; !exist {
+		klog.V(5).Infof("No individual jobInfo waiting time settings for jobInfo <%s/%s>.", jobInfo.Namespace, jobInfo.Name)
+		// if no individual settings, read global jobInfo waiting time from sla plugin arguments
+		return sp.jobWaitingTime, nil
```
Comment (k82cn) [inline pkg/scheduler/plugins/sla/sla.go:58]: if no waiting time, why return nil as error?

File: pkg/scheduler/plugins/sla/sla.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+	jobWaitingTime, err := time.ParseDuration(jobInfo.PodGroup.Annotations[JobWaitingTime])
+	if err != nil {
+		klog.Errorf("Error occurs in parsing jobInfo waiting time for jobInfo <%s/%s>, err: %s.",
+			jobInfo.Namespace, jobInfo.Name, err.Error())
+		return time.Duration(-1), err
+	}
+	if jobWaitingTime < 0 {
+		klog.Warningf("Invalid individual jobInfo waiting time settings: %f minutes for jobInfo <%s/%s>.",
+			jobWaitingTime.Minutes(), jobInfo.Namespace, jobInfo.Name)
+		return time.Duration(-1), nil
```
Comment (k82cn) [inline pkg/scheduler/plugins/sla/sla.go:70]: ditto

File: pkg/scheduler/api/job_info.go [hunk 8]
```diff
@@ -140,6 +145,8 @@ type JobInfo struct {
 
 	MinAvailable int32
 
+	SLAWaitingTime *time.Duration
```
Comment (k82cn) [inline pkg/scheduler/api/job_info.go:148]: s/SLAWaitingTime/WaitingTime

File: pkg/scheduler/api/job_info.go [hunk 9]
```diff
@@ -199,6 +207,28 @@ func (ji *JobInfo) SetPodGroup(pg *PodGroup) {
 
 }
 
+// GetSLAWaitingTime reads sla waiting time for job from podgroup annotations
+// TODO: should also read from given field in volcano job spec
+func GetSLAWaitingTime(pg *PodGroup) *time.Duration {
```
Comment (k82cn) [inline pkg/scheduler/api/job_info.go:212]: return (*time.Duration, error), and did not print warning log

File: pkg/scheduler/plugins/util/util.go [hunk 10]
```diff
@@ -28,6 +28,15 @@ import (
 	"volcano.sh/volcano/pkg/scheduler/framework"
 )
 
+const (
+	// Permit indicates that plugin function permit this job to be inqueue, pipelined, or other status
+	Permit = 1
+	// Permit indicates that plugin function abstain in voting this job to be inqueue, pipelined, or other status
```
Comment (k82cn) [inline pkg/scheduler/plugins/util/util.go:34]: s/Permit/Abstain

File: pkg/scheduler/plugins/util/util.go [hunk 11]
```diff
@@ -28,6 +28,15 @@ import (
...
+	Abstain = 0
+	// Permit indicates that plugin function reject this job to be inqueue, pipelined, or other status
```
Comment (k82cn) [inline pkg/scheduler/plugins/util/util.go:36]: s/Permit/Reject

File: pkg/scheduler/plugins/sla/sla.go [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+	// read in sla waiting time for global cluster from sla plugin arguments
+	// if not set, job waiting time still can set in job yaml separately, otherwise job have no sla limits
+	if _, exist := sp.pluginArguments[JobWaitingTime]; exist {
+		jwt, err := time.ParseDuration(sp.pluginArguments[JobWaitingTime])
+		if err != nil {
+			klog.Errorf("Error occurs in parsing global job waiting time in sla plugin, err: %s.", err.Error())
+		}
+
+		if jwt < 0 {
+			klog.V(4).Infof("Invalid global sla waiting time settings in sla plugin: %f minutes.", jwt.Minutes())
```
Comment (william-wang) [inline pkg/scheduler/plugins/sla/sla.go:94]: please use warning for invalid configuration.

File: pkg/scheduler/plugins/sla/sla.go [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+		} else {
+			sp.jobWaitingTime = &jwt
+			klog.V(4).Infof("Global job waiting time is %f minutes.", sp.jobWaitingTime.Minutes())
```
Comment (william-wang) [inline pkg/scheduler/plugins/sla/sla.go:97]: No need print info log for each job. It cause too much log.

**Comment by k82cn** [comment]
LGTM overall :)

**Review by william-wang — APPROVED** [review]
/lgtm
/approve

File: pkg/scheduler/plugins/sla/sla.go (hunk 4 above)
Comment (jiangkaihua) [reply pkg/scheduler/plugins/sla/sla.go:2]: already updated.

**Pull request merged**